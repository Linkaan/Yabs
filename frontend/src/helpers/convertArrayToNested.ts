export default function convertList(payload: any[], typeOfId: string): object { //eslint-disable-line
  return payload.reduce((acc, item) => {
    acc[item[typeOfId]] = item;
    return acc;
  }, {});
}
