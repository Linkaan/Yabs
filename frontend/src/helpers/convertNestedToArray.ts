export default function convertList(payload: any, typeOfId: string): any[] { //eslint-disable-line
  const res = Object.keys(payload).map((id) => payload[id]);
  return res;
}
