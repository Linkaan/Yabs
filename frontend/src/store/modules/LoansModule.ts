import Vue from 'vue';
import {VuexModule, Module, getModule, Action, Mutation} from 'vuex-module-decorators';
import store from '..';
import LoansAPI from '../../api/loans';
import convertList from '@/helpers/convertArrayToNested';
import convertNested from '@/helpers/convertNestedToArray';

export interface Loan {
  book: object;
  book_id: number; //eslint-disable-line camelcase
  created_at: string; //eslint-disable-line camelcase
  expiration_date: string; //eslint-disable-line camelcase
  id: number;
  lent_by: object; //eslint-disable-line camelcase
  lent_by_id: number; //eslint-disable-line camelcase
  loaned_by: object; //eslint-disable-line camelcase
  loaned_by_id: number; //eslint-disable-line camelcase
  returned_at: string|null; //eslint-disable-line camelcase
  updated_at: string; //eslint-disable-line camelcase
}

export interface LoanForm {
  id?: number;
  book_id: number; //eslint-disable-line camelcase
  lent_by_id: number; //eslint-disable-line camelcase
  loaned_by_id: number; //eslint-disable-line camelcase
}

export interface LoanCollection {[id: number]: Loan; }

@Module({dynamic: true, namespaced: true, name: 'LoansModule', store})
class LoansModule extends VuexModule {
  private _loans: LoanCollection = {};
  private _failure: object = {};

  get all(): LoanCollection {
    return this._loans;
  }

  get allAsArray(): Loan[] {
    return convertNested(this._loans);
  }

  @Action({rawError: true})
  public fetchAll(): Promise<LoanCollection> {
    return new Promise((resolve, reject) => {
      LoansAPI.all()
        .then((response: Loan[]) => {
          this.convertLoanList(response);
          resolve(this._loans);
        })
        .catch((error: object) => {
          reject(error);
        });
    });
  }

  @Action({rawError: true})
  public create(request: LoanForm): Promise<Loan> {
    return new Promise((resolve, reject) => {
      LoansAPI.create(request)
        .then((response: Loan) => {
          this.setLoan(response);
          resolve(response);
        })
        .catch((error: object) => {
          reject(error);
        });
    });
  }

  @Action({rawError: true})
  public update(request: LoanForm): Promise<Loan> {
    return new Promise((resolve, reject) => {
      LoansAPI.update(request)
        .then((response: Loan) => {
          this.setLoan(response);
          resolve(response);
        })
        .catch((error: object) => {
          reject(error);
        });
    });
  }

  @Action({rawError: true})
  public delete(request: Loan): Promise<number> {
    return new Promise((resolve, reject) => {
      LoansAPI.delete(request)
        .then((response: number) => {
          this.removeLoan(response);
          resolve(response);
        })
        .catch((error: object) => {
          reject(error);
        });
    });
  }

  @Mutation
  private setLoan(payload: Loan):void {
    Vue.set(this._loans, payload.id, payload);
  }

  @Mutation
  private removeLoan(loanId: number): void {
    Vue.delete(this._loans, loanId);
  }

  @Mutation
  private convertLoanList(payload: Loan[]): void {
    const list = convertList(payload, 'id');
    this._loans = list;
  }

  @Mutation
  private setFailure(payload: object): void {
    //should enable e2e CI pipeline
  }
}

export default getModule(LoansModule);
