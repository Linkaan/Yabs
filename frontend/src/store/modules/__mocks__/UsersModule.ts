import { User } from '@/types';

const user: User = {
  'uid': 123,
  'name': 'Example',
  'email': 'example@example.example',
  'klass': 'unknown',
  'role': 1,
  'google_token': '11111111111111111',
  'photo_path': 'abc.web',
  'created_at': '2020-12-21T15:22:58.998Z',
  'updated_at': '2020-12-21T15:22:58.998Z'
};
const user1: User = {
  'uid': 444,
  'name': 'Test',
  'email': 'test@test.test',
  'klass': 'test',
  'role': 20,
  'google_token': '11111111111111111',
  'photo_path': 'abc.web',
  'created_at': '2020-12-21T15:22:58.998Z',
  'updated_at': '2020-12-21T15:22:58.998Z'
};

module.exports = {
  allAsArray: [user, user1],
  fetchAll: () => null,
  create: jest.fn().mockImplementationOnce(_arg => {
    return Promise.resolve();
  }),
  update: jest.fn(),
  fetchSingle: (_: any) => {
    return user;
  },
};
