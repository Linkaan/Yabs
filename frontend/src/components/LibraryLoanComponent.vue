<template>
  <div>
    <v-card data-cy="libraryLoanCard">
      <v-card-title>
        Skanna bokens streckkod
      </v-card-title>
      <v-form
        v-if="show"
        @submit="onSubmit"
        @reset="onReset"
      >
        <v-item-group
          vertical=""
          label-class="text-sm-right"
          label-for="nestedBid"
        >
          <v-text-field
            id="nestedBid"
            v-model="form.book_id"
            data-cy="book_barcode"
            label="Bokens Streckkod"
            outlined
          />
        </v-item-group>
        <v-card-actions>
          <v-spacer />
          <v-btn
            data-cy="loanOutBook"
            type="submit"
            color="primary"
            class="mr-4"
            large
          >
            LÃ¥na Ut
          </v-btn>
          <v-spacer />
        </v-card-actions>
      </v-form>
    </v-card>
  </div>
</template>

<script lang="ts">
import LoansModule from '../store/modules/LoansModule';
import { Loan, LoanForm } from '../types';
import UsersModule from '../store/modules/UsersModule';
import { ref, defineComponent, SetupContext } from '@vue/composition-api';
import LoanFormComponent from './LoanFormComponent.vue';

export default defineComponent ({
  name: 'LibraryLoanComponent',
  components: {
    LoanFormComponent
  },
  setup(_: object, { root, emit }: SetupContext) {
    const form = ref({
      lent_by_id: 0, //eslint-disable-line camelcase
      loaned_by_id: 0, //eslint-disable-line camelcase
      book_id: 0, //eslint-disable-line camelcase
    } as LoanForm);

    const show = ref(true);

    function onSubmit(evt: Event): void {
      evt.preventDefault();
      form.value.lent_by_id = 2054282603; //eslint-disable-line camelcase
      form.value.loaned_by_id = UsersModule.currentUserID; //eslint-disable-line camelcase
      if (!!form.value.lent_by_id && !!form.value.loaned_by_id && !!form.value.book_id) {
        LoansModule.create(form.value)
          .then((payload: Loan) => emit('loan-added', payload))
          .catch((failure: boolean) => console.log(failure));
      }
    }

    function onReset(evt: Event): void {
      evt.preventDefault();
      form.value.loaned_by_id = 0; //eslint-disable-line camelcase
      form.value.book_id = 0; //eslint-disable-line camelcase
      show.value = false;
      root.$nextTick(() => {
        show.value = true;
      });
    }
    return { form, show, LoanFormComponent, onSubmit, onReset };
  }
});
</script>
