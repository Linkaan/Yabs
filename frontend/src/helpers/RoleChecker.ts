import UsersModule, { User } from '../store/modules/UsersModule';

export default class RoleChecker {

  private static checkPermission(user: User, required: number): boolean {
    return user && ((user.role & required) == required || (user.role > required));
  }
  
  static isPrincipal(): boolean {
    return this.checkPermission(UsersModule.currentUser, 32);
  }

  static isAdmin(): boolean {
    return this.checkPermission(UsersModule.currentUser, 16);
  }

  static isCaretaker(): boolean {
    return this.checkPermission(UsersModule.currentUser, 8);
  }

  static isTeacher(): boolean {
    return this.checkPermission(UsersModule.currentUser, 4);
  }

  static isStudentHealth(): boolean {
    return this.checkPermission(UsersModule.currentUser, 2);
  }

  static isStudent(): boolean {
    return UsersModule.currentUser && (UsersModule.currentUser.role & 1) == 1;
  }
}
