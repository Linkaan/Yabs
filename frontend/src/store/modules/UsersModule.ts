import Vue from 'vue';
import {
  VuexModule,
  Module,
  getModule,
  Action,
  Mutation,
} from 'vuex-module-decorators';
import store from '..';
import UsersAPI from '../../api/users';
import convertList from '../../helpers/convertArrayToNested';
import RoleChecker from '@/helpers/roleChecker';

export interface User {
  created_at: string; //eslint-disable-line camelcase
  email: string;
  google_token: string; //eslint-disable-line camelcase
  klass: string;
  name: string;
  photo_path: string; //eslint-disable-line camelcase
  role: number;
  uid: number;
  updated_at: string; //eslint-disable-line camelcase
}

export interface UserCollection { [uid: number]: User; }

@Module({dynamic: true, namespaced: true, name: 'UsersModule', store})
class UsersModule extends VuexModule {
  private _failure: object = {};
  public _users: UserCollection = {};
  private _currentUser: number = 0;

  get currentUserID(): number {
    return this._currentUser;
  }

  get all(): UserCollection {
    return this._users;
  }

  get currentUser(): User {
    return this._users[this._currentUser];
  }

  get roleAsText(): string {
    if (!this.currentUser) {return ''; }
    let output: string = '';
    if (RoleChecker.isStudent()) {output = 'Elev'; 
    } else if (RoleChecker.isPrincipal()) {output = 'Rektor';
    } else if (RoleChecker.isAdmin()) {output = 'Administratör';
    } else if (RoleChecker.isCaretaker()) {output = 'Vaktmästare';
    } else if (RoleChecker.isTeacher()) {output = 'Lärare';
    } else if (RoleChecker.isStudentHealth()) {output = 'Elevhälsoteam'; }
    return output;
  }

  @Action({rawError: true})
  public fetchAll(): Promise<object> {
    return new Promise((resolve, reject) => {
      UsersAPI.all()
        .then((response: User[]) => {
          this.convertUserList(response);
          resolve(this._users);
        })
        .catch((error: object) => {
          this.setFailure(error);
          reject(error);
        });
    });
  }

  @Action({rawError: true})
  public update(request: object): Promise<object> {
    return new Promise((resolve, reject) => {
      UsersAPI.update(request)
        .then((response: User) => {
          this.setUser(response);
          this.setCurrentUser(response);
          resolve(response);
        })
        .catch((error: object) => {
          reject(error);
        });
    });
  }

  @Action({rawError: true})
  public signIn(request: string): Promise<object> {
    return new Promise((resolve, reject) => {
      UsersAPI.signIn(request)
        .then((response: User) => {
          this.setCurrentUser(response);
          this.fetchAll();
          resolve(response);
        })
        .catch((error: object) => {
          this.setFailure(error);
          reject(error);
        });
    });
  }

  @Action({rawError: true})
  public signOut(): Promise<object> {
    return new Promise((resolve, reject) => {
      UsersAPI.signOut()
        .then((response: User) => {
          this.setCurrentUser(response);
          resolve(response);
        })
        .catch((error: object) => {
          reject(error);
        });
    });
  }

  @Mutation
  public setUser(payload: User): void {
    Vue.set(this._users, payload.uid, payload);
  }

  @Mutation
  public setCurrentUser(payload: User): void {
    this._currentUser = payload.uid;
  }

  @Mutation
  public setFailure(payload: object): void {
    console.log(payload);
    // This should fix end to end tests(yarn build)
  }

  @Mutation
  public convertUserList(payload: User[]): void {
    const list = convertList(payload, 'uid');
    this._users = list;
  }
}

export default getModule(UsersModule);
