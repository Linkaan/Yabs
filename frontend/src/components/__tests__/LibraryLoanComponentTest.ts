import LibraryLoanComponent from '@/components/LoanFormComponent.vue';

import { factory } from '@/helpers/testFactoryHelpers';
import LoansModule from '@/store/modules/LoansModule';
import UsersModule from '@/store/modules/UsersModule';
import { User } from '@/types';

// @ts-ignore

jest.mock('@/store/modules/LoansModule', () => {
  return {
    create: jest.fn().mockImplementationOnce((_arg) => {
      return Promise.resolve();
    })
  };
});

describe('LoanFormComponent.vue', () => {

  
    it('calls loansmodule with correct values and resets the input fields', async () => {
      const wrapper = factory(LibraryLoanComponent);
      UsersModule.setCurrentUser({uid: 123} as User);
      console.log(UsersModule.currentUserID);
  
      wrapper.find('[data-jest=\'book_barcode\']').setValue('1234');
      await wrapper.vm.$nextTick();
      wrapper.find('[data-jest=\'form\']').trigger('submit');
      await wrapper.vm.$nextTick()
  
      expect(LoansModule.create).toHaveBeenCalledWith({
        loaned_by_id: '123', //eslint-disable-line camelcase
        book_id: '1234', //eslint-disable-line camelcase
        lent_by_id: 123, //eslint-disable-line camelcasee
      });
    });
  });