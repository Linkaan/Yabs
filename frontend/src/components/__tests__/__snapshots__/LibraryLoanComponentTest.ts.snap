// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoanFormComponent.vue renders correctly 1`] = `
<div>
  <v-btn-stub tag="button" activeclass="" type="button" data-jest="barcodeButton">
    <v-icon-stub arialabel="L채s av Streckkoder" aria-label="L채s av Streckkoder">
      mdi-barcode-scan
    </v-icon-stub>
  </v-btn-stub>
  <v-form-stub data-jest="form">
    <v-item-group-stub activeclass="v-item--active" vertical="" label-class="text-sm-right" label-for="nestedUid">
      <v-text-field-stub errorcount="1" errormessages="" messages="" rules="" successmessages="" value="0" backgroundcolor="" id="nestedUid" label="Elevens Streckkod" loaderheight="2" clearicon="$clear" outlined="true" type="text" data-jest="student_barcode" data-cy="student_barcode"></v-text-field-stub>
    </v-item-group-stub>
    <v-item-group-stub activeclass="v-item--active" vertical="" label-class="text-sm-right" label-for="nestedBid">
      <v-text-field-stub errorcount="1" errormessages="" messages="" rules="" successmessages="" value="0" backgroundcolor="" id="nestedBid" label="Bokens Streckkod" loaderheight="2" clearicon="$clear" outlined="true" type="text" data-jest="book_barcode" data-cy="book_barcode"></v-text-field-stub>
    </v-item-group-stub>
    <v-item-group-stub activeclass="v-item--active" vertical="" label-class="text-sm-right" label-for="nestedBid">
      <v-menu-stub opendelay="0" closedelay="0" contentclass="" maxwidth="auto" minwidth="290px" nudgebottom="0" nudgeleft="0" nudgeright="0" nudgetop="0" nudgewidth="0" openonclick="true" closeonclick="true" maxheight="auto" offsety="true" origin="top left" transition="scale-transition">
        <v-date-picker-stub width="290" eventcolor="warning" firstdayofweek="0" min="2021-01-15" nexticon="$next" previcon="$prev" showcurrent="true" selecteditemstext="$vuetify.datePicker.itemsSelected" type="date" value=""></v-date-picker-stub>
      </v-menu-stub>
    </v-item-group-stub>
    <v-btn-stub color="primary" tag="button" activeclass="" large="true" type="submit" data-cy="loanOutBook" class="mr-4">
      L책na Ut
    </v-btn-stub>
    <v-btn-stub tag="button" activeclass="" large="true" type="reset">
      Rensa F채lten
    </v-btn-stub>
  </v-form-stub>
</div>
`;
